posts/2018-02-09-serverless-and-graphql-in-love.md

```diff
@@ -0,0 +1,554 @@
+
+[GraphQL](https://dev-blog.apollodata.com/2017-the-year-in-graphql-124a050d04c6) lets you shrink your multitude of APIs down into a single HTTP endpoint, which you can use to fetch data from multiple data sources. In short, it provides a simple way of building mobile/web applications by providing a clean layer of abstraction between servers and clients.
+
+It lets you:
+1. Reduce network costs and get better query efficiency.
+2. Know exactly what your response will look like and ensure you're never sending more or less than the client needs.
+3. Describe your API with types that map your schema to existing backend.
+
+Thousands of companies are now using GraphQL in production with the help of open source frameworks built by Facebook, Apollo, and Graphcool. In fact, that [Starbucks announcement](https://twitter.com/davidbrunelle/status/960946257643454464) last week made my morning coffee taste even better. 😉
+
+![](https://user-images.githubusercontent.com/1587005/36067514-503a05b4-0e73-11e8-9b40-946c5398f4b5.png "Danielle's slide from Serverless and GraphQL meetup at Glassdoor, Jan 29, 2018")
+
+Now, very reasonably, you are probably thinking, “Yeah, okay, Facebook is one thing; they have a giant engineering team. But for me, having only one API endpoint is too risky. What if it goes down? How do I handle that much load? What about security?”
+
+You are absolutely correct: with one HTTP endpoint, you need be entirely sure that endpoint never goes down and that it scales on demand.
+
+That’s where serverless comes in.
+
+## What is Serverless?
+
+Serverless has gained popularity over last few years, primarily because it gives developers flexibility.
+
+With Serverless comes the following:
+1. No server management (no need to manage any form of machine)
+2. Pay-per-execution (never pay for idle)
+3. Authentication and authorization at scale
+4. Function as a unit of application logic
+
+## What makes Serverless and GraphQL such a great fit?
+
+When moving to GraphQL, you suddenly rely on one HTTP endpoint to connect your clients to your backend services. Once you do decide to do that, you want this one HTTP endpoint to be: reliable, fast, auto-scaling, and have a small attack vector regarding security.
+
+*All* these properties are fulfilled by a single AWS Lambda function in combination with API Gateway. It’s just a great fit! 
+
+**Note:** Check out [Jared Short's Serverless + GraphQL post](https://www.trek10.com/blog/a-look-at-serverless-graphql/) to get an even better understanding of this relationship 😎
+
+In sum, powering your GraphQL endpoint with a serverless backend solves scaling and availability concerns outright, and it gives you a big leg up on security. It’s not even that much code or configuration.
+
+It takes only a few minutes to get to a production-ready setup, which we're about to dive into, right now.
+
+## Serverless-GraphQL repository
```

Comment:
call out monitoring, logging - but you could use apollo server hapi and apollo server express can be used as well. 
lambda (abstraction over express, removes complexities of express) vs hapi vs express (more compilcated to setup).

Reply:
Here's something you can use to start with:

In this example, we're using [`apollo-server-lambda`](https://github.com/apollographql/apollo-server/tree/master/packages/apollo-server-lambda), which is a Node.js server library which is designed to quickly get you up and running on AWS Lambda. Now you might be more familiar with server frameworks such as Express, Koa, or Hapi, and there are Apollo Server plugins for all of these frameworks. You can use any of these with Serverless as well, but they each require a little more configuration which is beyond the scope of this blog post.

Perhaps might want to mention in a follow-up post (FAQ) what the trade-offs are between using something like `apollo-server-lambda` and `apollo-server-express`. The same can be said for the trade-offs between using Serverless vs more traditional server environments such as a Docker container. One that I've discovered is that you can't use some monitoring solutions with Serverless, such as OpBeat or Sentry, which bucket and send off performance monitoring data at set intervals, which require a long-running process, which isn't how Lambda operates.

Reply:
done